### DISCUSSION TITLE
아이템 68. 일반적으로 통용되는 명명 규칙을 따르라

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Effective-Java/Effective-Java#158] [3. 과제]
## 핵심 요약

- 표준 명명 규칙을 체화하여 자연스럽게 배어 나오도록 하자
- 자바 언어 명세(_Java Language Specification_, [_JLS_, 6.1](https://docs.oracle.com/javase/specs/jls/se7/html/jls-6.html#jls-6.1))에 다음과 같은 인용 구절이 존재
  > "오랫동안 따라온 규칙과 충돌한다면 그 규칙을 맹종해서는 안 된다."
- 자바의 명명 규칙은 패키지, 클래스, 인터페이스, 필드, 메서드 등에 대해 크게 철자와 문법, 두 범주로 구분

## 패키지

- 패키지의 각 요소는 일반적으로 8자 이하의 짧은 단어로 구성
  (e.g. `utilities`* -> `util`)
- 여러 단어로 구성된 이름이라면 각 단어의 첫 글자만 따서 사용 가능
  (e.g. `aop`)
- 많은 기능을 제공하는 경우 계층을 나눠 더 많은 요소로 구성 가능
  이처럼 자바 패키지 계층에서 하부 패키지를 하위 패키지(*subpackage*)라고 함
  (e.g. `java.util.concurrent.atomic`)

## 클래스, 인터페이스

- 열거 타입과 애너테이션을 포함한 클래스와 인터페이스의 이름은 파스칼 표기법을 따름
- 여러 단어의 첫 글자만 딴 약자나 `max`, `min` 처럼 널리 통용되는 줄임말을 제외하면 단어 줄이기 금지
- 약자의 경우 첫글자만 대문자로 할지 전체를 대문자로 할지는 논란이 있음 :star:
  (e.g. `NHNCloudService` vs. `NhnCloudService`)
- 열거 타입을 포함한 객체를 생성할 수 있는 클래스 이름은 보통 단수 명사나 명사구 사용 :star:
  (e.g. `Thread`, `PriorityQueue`)
- 객체를 생성할 수 없는 클래스 이름은 보통 복수형 명사 사용 :star:
  (e.g. `Collections`, `StringUtils`)
- 인터페이스 이름은 클래스와 똑같이 짓거나 `-able`, `-ible` 처럼 형용사형 어미 사용
  - (e.g. `UserDetailsService`, `AutoCloseable`, `Serializable`)

## 필드

- 상수 필드를 구성하는 단어는 모두 대문자로 쓰며, 단어 사이는 밑줄로 구분
  (e.g. `INSTANCE`, `MAX_VALUE`) (feat. #37)

## 메서드

- 어떤 동작을 수행하는 메서드의 이름은 동사나 목적어를 포함한 동사구로 명명
  (e.g. `append`, `drawImage`)
- `boolean` 값을 반환하는 메서드는 보통 `is-`, `has-` 접두어 사용
  (e.g. `isEmpty`, `hasSiblings`)
- 객체의 타입을 바꿔 다른 타입의 또 다른 객체를 반환하는 인스턴스 메서드 이름은 보통 `toType` 형태로 명명
  (e.g. `toString`, `toArray`)
- 객체의 내용을 다른 뷰로 보여주는 메서드의 이름은 `asType` 형태로 명명
  (e.g. `asList`) (feat. #11)

## 타입 매개변수

- 보통 한 문자로 표현
  - `T`: 임의의 타입
  - `E`: 컬렉션 원소의 타입
  - `K`, `V`: 맵의 키와 값
  - `X`: 예외
  - `R`: 메서드의 반환 타입
  - `T`, `U`, `V` or `T1`, `T2`, `T3`: 임의 타입의 시퀀스

## 회고

예전만큼은 아니지만 평소에도 해당 주제에 대해 잘 지켜야 한다는 고집이 있는데, 잘 알려진 도서에서 가이드를 제시해주니 제 마음이 마치 아래와 같았습니다.

| ![gaviscon-incomfortable](https://user-images.githubusercontent.com/38161720/226161474-603eed52-5284-4162-8de9-54866b425a79.png) | ![gaviscon-comfortable](https://user-images.githubusercontent.com/38161720/226161484-66d91c16-9eb2-44b4-9065-10add6e96380.png) |
|----------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------|

**Comment by Irisation23** [comment]
앞으로 네이밍 룰을 적용해서 직접적으로 프로젝트에 참여하게 될 상황이 생길 것 같은데
해당 주제 필두로 프로젝트에 임하도록 해보겠습니다.

정리 감사합니다. 😁

**Comment by YuDeokRin** [comment]
네이밍 규칙을 구체적으로 몰랐는데... 이 글에서 많은걸 얻어갑니다..
